Iterable

smichr opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->

spinoff from #22582 with some additional changes

#### Brief description of what is fixed or changed

1. `Expr.__eq__` automatically rejects as unequal anything that is iterable (and also saves time by not having to sympify anything in the iterable
2. The default arg z=None is handled without putting it into a Tuple
3. it looked like NotIterable was in campatibililty at one time -- the tests for compatibility wanted to import it from there -- so it was added along with the others that are deprecated in favor of importing them from iterables.

#### Other comments


#### Release Notes

<!-- Write the release notes for this release below between the BEGIN and END
statements. The basic format is a bulleted list with the name of the subpackage
and the release note for this PR. For example:

* solvers
  * Added a new solver for logarithmic equations.

* functions
  * Fixed a bug with log of integers.

or if no release note(s) should be included use:

NO ENTRY

See https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more
information on how to write release notes. The bot will check your release
notes automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
NO ENTRY
<!-- END RELEASE NOTES -->

smichr committed:
NotIterable has new home

smichr committed:
iterable != Expr

There is no need to sympify our way through a tuple
to find that the tuple is not equal to
an expression. So this is a quick exit.

smichr committed:
don't put None in Tuple

ThePauliPrinciple commented on sympy/core/random.py:
```diff
@@ -72,10 +72,13 @@ def verify_numerically(f, g, z=None, tol=1.0e-6, a=2, b=-1, c=3, d=1):
-    from sympy.core.containers import Tuple
-    f, g, z = Tuple(f, g, z)
-    z = [z] if isinstance(z, Symbol) else (f.free_symbols | g.free_symbols)
+    f, g = (sympify(i) for i in (f, g))
```
There is a similar (mis)use of `Tuple` here: https://github.com/sympy/sympy/blob/ed0bfd1e47345c8bd45ef542eef18f6a21c55685/sympy/integrals/integrals.py#L808-L813

ThePauliPrinciple replied:
And here:
https://github.com/sympy/sympy/blob/ed0bfd1e47345c8bd45ef542eef18f6a21c55685/sympy/tensor/indexed.py#L326-L334

ThePauliPrinciple commented on sympy/core/expr.py:
```diff
@@ -146,11 +146,15 @@ def _hashable_content(self):
-        try:
-            other = _sympify(other)
-            if not isinstance(other, Expr):
+        if not isinstance(other, Basic):
```
Note that `Tuple` defines its own custom `__eq__`: https://github.com/sympy/sympy/blob/ed0bfd1e47345c8bd45ef542eef18f6a21c55685/sympy/core/containers.py#L96-L99

Although I do believe this is handled correctly.

ThePauliPrinciple commented:
Looks good.

If it is not possible to solve the two other uses of `Tuple` (provided in the review) I think this can be accepted.

ThePauliPrinciple commented on sympy/core/expr.py:
```diff
@@ -146,11 +146,15 @@ def _hashable_content(self):
-            other = _sympify(other)
-            if not isinstance(other, Expr):
+        if not isinstance(other, Basic):
+            if iterable(other):
```
looks like `dict` is not considered an `iterable`, but I think it should get a similar treatment

smichr replied:
Dict is Basic but not Expr so it will get reject without sympification.

ThePauliPrinciple replied:
I mean a python `dict`, which is not a `Basic`, nor will it go to the `iterable` branch, so it will try to `sympify` the `dict`, potentially creating `Dict` with `None` just like it did with `tuple->Tuple` before

oscarbenjamin commented:
ImmutableMatrix is one example of an iterable Expr. So is MatrixSymbol but iterable thinks that it isn't:
```python
In [4]: from sympy.utilities.iterables import iterable

In [5]: iterable(ImmutableMatrix([[1, 2]]))
Out[5]: True

In [7]: x = MatrixSymbol('x', 2, 2)

In [8]: iterable(x)
Out[8]: False

In [9]: list(x)
Out[9]: [x₀₀, x₀₁, x₁₀, x₁₁]
```

ThePauliPrinciple commented:
But `MatrixSymbol` is a `Basic`. So it shouldn't matter here since `iterable` is only used for non-`Basic` equalities.

oscarbenjamin commented:
Yes, but the question is whether an Expr could ever be considered equal to a non-Basic iterable e.g. something like:
```python
In [6]: import sympy as sym

In [7]: import numpy as np

In [8]: sym.Matrix([[1, 2], [3, 4]]) == np.array([[1, 2], [3, 4]])
Out[8]: False
```
I don't know of any examples but I'm just considering it.

ThePauliPrinciple commented:
`Tuple(1, 2, 3) == (1, 3, 4)`

Note that `Matrix` is not an `Expr`.

oscarbenjamin commented:
> Note that `Matrix` is not an `Expr`.

ImmutableMatrix is though.

ThePauliPrinciple commented:
```python
>>> from sympy import ImmutableMatrix
>>> import numpy as np
>>> print(ImmutableMatrix([[1, 2], [3, 4]]) == np.array([[1, 2], [3, 4]]))
None
```
Looks like a bug (on 1.9)

smichr committed:
remove (ab)use of Tuple

smichr commented:
> Looks like a bug (on 1.9)

Since `_sympify` is strict, the np array is not converted.

```python
>>> a=(np.array([[1, 2], [3, 4]]))
>>> s=S(a); type(s)
<class 'sympy.tensor.array.dense_ndim_array.ImmutableDenseNDimArray'>
>>> s==a
False
```

Also, the result is not a Matrix (and not Expr) so the changed code does not apply in this case.

But if something does define `_symy_` is should be handled. I made a change towards that end.

smichr committed:
watch for iterable Expr

Each iterable class needs its own __eq__ method.
Thus guarded, Expr.__eq__ can reject (and often does)
anything that is iterable without the need to look
further.

ThePauliPrinciple commented:
Yes, it isn't caused due to this PR, but surely `==` should not return `None`.

smichr commented:
I'm happy with how often this shortcut actually gets used in the tests.

smichr committed:
check for _sympy_ on other iterable

smichr commented:
This now only returns a False if the iterable other doesn't have a `_sympy_` method, so I think it addresses the concern of @oscarbenjamin . I think this is ready if this run passes.

ThePauliPrinciple commented:
Does it work yet for `dict`? It doesn't pass `iterable({})` but I think it should be treated in the same way

oscarbenjamin commented on sympy/core/expr.py:
```diff
@@ -146,11 +146,17 @@ def _hashable_content(self):
+                # XXX iterable self should have it's own __eq__
+                # method if the path gives a false negative
+                # comparison
+                return False
```
This should be `return NotImplemented`.

smichr replied:
done

smichr committed:
return NotImplemented for unrecognized obj

ThePauliPrinciple commented on sympy/core/expr.py:
```diff
@@ -146,11 +146,17 @@ def _hashable_content(self):
-            other = _sympify(other)
-            if not isinstance(other, Expr):
+        if not isinstance(other, Basic):
+            if iterable(other) and not hasattr(other, '_sympy_'):
```
```python
if (iterable(other) or isinstance(other, dict)) and not hasattr(other, '_sympy_')
```

ThePauliPrinciple replied:
preferable : 
```python
if iterable(other, exclude = (str, NotIterable)) and not hasattr(other, `_sympy_`):
```
Looks like `dict` is excluded by default

smichr committed:
reject dict, too

ThePauliPrinciple commented:
Thanks a lot! I believe this will solve a lot of the remaining errors in https://github.com/sympy/sympy/pull/22445 and would allow for https://github.com/sympy/sympy/pull/22580 to work.

smichr commented:
> Thanks a lot! I believe this will solve a lot of the remaining errors 

Thanks for your (and @oscarbenjamin ) feedback.
